### ISSUE TITLE
needs_warnings not returning error code or writing to error log

### ISSUE THREAD

**Issue by twodrops** [useblocks/sphinxcontrib-needs#344]
Usage of ` -W` shall handle all warnings as errors and the sphinx-build executable shall return a non-zero value as error code. Also the usage of `-w <file>` shall write all warnings to a file.

This works incase of Sphinx warnings, but not for warnings created with `needs_warnings`. Because of this pipelines cannot be configured to break incase of `needs_warnings`.

To reproduce the problem, do the following steps

1. Create `needs_warnings` in `conf.py`
2. Create a need warning within any rST
3. Execute sphinx-build as follows from a console
```bash
sphinx-build -M html docs _build -W --keep-going -w error.log  
```
4. Observe `need_warnings` on the console
5. Execute the following from the console
```bash
echo $?
```
6. `0` is returned (a non-zero value is expected)
7. Check the file `error.log`
8. Only the following line is present
```
WARNING: Sphinx-Needs warnings were raised. See console / log output for details.
```
All `needs_warnings` are expected in the log.

Could it be that `needs_warnings` is always writing to stdout instead of stderror (even with -W)?

**Comment by danwos** [comment]
Thanks for the great bug reports. It helps a lot.

To be honest, I thought ``-W`` gets handled the correct way, as I use it in some project as well.

It is true, that the output of the single warning checks are producing a normal ``log.info``on stdout.
But if one or more warnings are detected, a final ``log.warning`` is written, with the text:
"Sphinx-Needs warnings were raised. See console / log output for details."
This final ``log.warning`` should then stop the build, if ``-W`` option is used.

The used logger is coming from sphinx.

Will check, if raising an exception or even a specific ``sphinx.errors.SphinxWarning`` will set the exit code more stable.

**Comment by twodrops** [comment]
@danwos 

You are right. I observed too that the below statement would be present in the logfile
```Sphinx-Needs warnings were raised. See console / log output for details.```

What would be nice if all `need_warnings` also appear in the log. This could be a second step, if technical possible.

**Comment by danwos** [comment]
``--keep-going`` will always set exit code to 0, even if there are warnings.
This is based on sphinx and we can't do anything here.

For the warning message to error log: We can have a config option ``needs_warnings_always_warn``.
If ``True``, Sphinx-needs will raise a sphinx-warning for each not passed warning check (Default is ``False``).

So the first not passed warning will stop the build if used together with ``-W``.
If used with ``keep-going``, all warnings should get written to error log.

**Comment by twodrops** [comment]
@danwos @haiyangToAI  Thanks for the analysis on exit code and the feature for writing warnings to error log.

@danwos  I wanted to verify again the part about exit code which you mentioned.
>--keep-going will always set exit code to 0, even if there are warnings.

The sphinx documentation mentions this however differently (see below or [here](https://www.sphinx-doc.org/en/master/man/sphinx-build.html)). Here is it mentioned that an exit status of 1 will be raised. Is this then a Sphinx bug?

> -W
Turn warnings into errors. This means that the build stops at the first warning and sphinx-build exits with exit status 1.
> --keep-going
With -W option, keep going processing when getting warnings to the end of build, and sphinx-build exits with exit status 1.

**Comment by twodrops** [comment]
@haiyangToAI I tried to look into `tests/doc_test/doc_needlist` and `tests/doc_test/doc_needs_warnings`, but the difference between them was not directly clear. Could you please explain what the difference is, and what leads to exit code of 0 with the second test above.
